File: apps/patter-voice-sdk/libraries/typescript/src/observability/attributes.ts
/**
 * ``patter.*`` span attribute helpers — TypeScript mirror of
 * ``libraries/python/getpatter/observability/attributes.py``.
 *
 * No-op when OpenTelemetry isn't wired up. The helpers exist primarily so
 * provider adapters in both SDKs can share the same call sites without
 * each one re-implementing the "is OTel available?" gating. When tracing
 * is disabled (``PATTER_OTEL_ENABLED`` unset, ``@opentelemetry/api`` not
 * installed, or no active call scope), every helper is a fast no-op.
 *
 * Parity contract with the Python helpers:
 *   - ``record_patter_attrs``     ↔ ``recordPatterAttrs``
 *   - ``patter_call_scope``       ↔ ``patterCallScope``
 *   - ``attach_span_exporter``    ↔ ``attachSpanExporter``
 *
 * The semantics differ in one structural way: Python uses ``ContextVar``
 * to propagate the active call ID through asyncio task trees. JS has no
 * equivalent built-in, so this module uses a module-level "current call"
 * cell and a stack to support nested ``patterCallScope`` invocations on
 * the same loop. That's fine because the TS SDK has at most one active
 * call scope per Node process / per request handler in practice.
 */
import { getLogger } from '../logger';
import { ENV_FLAG, isTracingEnabled } from './tracing';

/**
 * "uut" = unit-under-test, the default value stamped on
 * ``patter.side`` when no driver/UUT split is configured.
 */
export const DEFAULT_SIDE = 'uut';

interface CallScopeFrame {
  readonly callId: string;
  readonly side: string;
}

const _scopeStack: CallScopeFrame[] = [];

function _currentScope(): CallScopeFrame | null {
  return _scopeStack.length > 0 ? _scopeStack[_scopeStack.length - 1] : null;
}

interface OtelApiShape {
  trace: {
    getTracer(name: string): {
      startSpan(
        name: string,
        options?: { attributes?: Record<string, unknown> },
      ): {
        setAttribute(key: string, value: unknown): unknown;
        end(): unknown;
        isRecording?(): boolean;
      };
    };
    getActiveSpan?(): {
      setAttribute(key: string, value: unknown): unknown;
      isRecording?(): boolean;
    } | null;
  };
}

function _tryLoadOtelApi(): OtelApiShape | null {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    return require('@opentelemetry/api') as OtelApiShape;
  } catch {
    return null;
  }
}

/**
 * Stamp ``patter.*`` attributes on the current span, augmenting them with
 * the ambient ``patter.call_id`` / ``patter.side`` from the active
 * ``patterCallScope``. No-op when tracing is disabled or no scope is
 * active.
 *
 * Behaviour mirrors the Python helper:
 *   - If an active recording span exists, attributes are stamped on it.
 *   - Otherwise a transient zero-duration ``patter.billable`` span is
 *     opened to carry the attributes. Some collectors filter
 *     zero-duration spans; callers that need guaranteed attribution
 *     should wrap their billable work in their own span.
 *
 * Caller-provided ``patter.call_id`` / ``patter.side`` keys win over the
 * scope's values.
 */
export function recordPatterAttrs(attrs: Readonly<Record<string, unknown>>): void {
  if (!isTracingEnabled()) return;
  const scope = _currentScope();
  if (scope === null) return;

  const api = _tryLoadOtelApi();
  if (!api) return;

  const full: Record<string, unknown> = { ...attrs };
  if (full['patter.call_id'] === undefined) full['patter.call_id'] = scope.callId;
  if (full['patter.side'] === undefined) full['patter.side'] = scope.side;

  try {
    const active = api.trace.getActiveSpan?.() ?? null;
    if (active && (active.isRecording === undefined || active.isRecording())) {
      for (const [k, v] of Object.entries(full)) {
        try {
          active.setAttribute(k, v);
        } catch {
          // Swallow — OTel must never crash the call path.
        }
      }
      return;
    }
  } catch {
    // fall through to billable-span fallback
  }

  try {
    const tracer = api.trace.getTracer('getpatter.observability');
    const span = tracer.startSpan('patter.billable', { attributes: full });
    try {
      span.end();
    } catch {
      // Swallow.
    }
  } catch {
    // Swallow.
  }
}

/**
 * Bind ``callId`` and ``side`` to the active span scope for the duration
 * of ``fn``. Mirrors the Python ``patter_call_scope`` context manager:
 * any ``recordPatterAttrs`` call made inside ``fn`` (or anything ``fn``
 * awaits) sees the bound values.
 *
 * Note: JavaScript has no ContextVar equivalent, so this uses a
 * module-level stack. Concurrent overlapping scopes on the same event
 * loop will see the innermost scope's values — fine for the SDK's
 * one-call-per-handler model. If callers need true async-context
 * isolation, install ``AsyncLocalStorage``-backed propagation via the
 * OTel SDK's context manager.
 */
export async function patterCallScope<T>(
  options: { readonly callId: string; readonly side?: string },
  fn: () => Promise<T>,
): Promise<T> {
  if (!options.callId) {
    throw new Error('patterCallScope requires non-empty callId');
  }
  const frame: CallScopeFrame = {
    callId: options.callId,
    side: options.side ?? DEFAULT_SIDE,
  };
  _scopeStack.push(frame);
  try {
    return await fn();
  } finally {
    // Defensive pop: locate this exact frame in case nested scopes
    // raised and left the stack in an unexpected state.
    const idx = _scopeStack.lastIndexOf(frame);
    if (idx >= 0) _scopeStack.splice(idx, 1);
  }
}

/**
 * Wire an OTel ``SpanExporter`` into the SDK's tracer provider and
 * remember the configured ``side`` on the Patter instance so the
 * per-call handler reads it when entering ``patterCallScope``.
 *
 * Mirrors the Python ``attach_span_exporter`` contract:
 *   - Stores ``side`` on ``patterInstance._patterSide`` unconditionally
 *     (works even when ``@opentelemetry/*`` peer deps are missing).
 *   - Idempotent on the *same exporter object reference*. Two distinct
 *     exporter instances pointing at the same backend will both be
 *     attached and spans will be exported twice — hold a single
 *     exporter object across calls to avoid duplicates.
 *
 * When tracing isn't enabled (env flag off / SDK peer deps absent), the
 * call is a no-op aside from storing ``_patterSide``.
 */
export function attachSpanExporter(
  patterInstance: { _patterSide?: string } & Record<string, unknown>,
  exporter: unknown,
  options: { readonly side?: string } = {},
): void {
  const side = options.side ?? DEFAULT_SIDE;
  patterInstance._patterSide = side;

  if (!isTracingEnabled()) {
    getLogger().debug(
      `attachSpanExporter: ${ENV_FLAG} not enabled or tracer unavailable; only side= stored`,
    );
    return;
  }

  // SDK wire-up packages are optional. When absent, fall through silently
  // — Python does the same.
  let sdkTraceBase:
    | {
        BasicTracerProvider: new (opts?: unknown) => {
          addSpanProcessor?(p: unknown): void;
          _patterAttachedExporters?: Set<unknown>;
        };
        SimpleSpanProcessor: new (exporter: unknown) => unknown;
      }
    | null = null;
  let sdkTraceNode:
    | {
        NodeTracerProvider: new (opts?: unknown) => {
          addSpanProcessor?(p: unknown): void;
          _patterAttachedExporters?: Set<unknown>;
        };
      }
    | null = null;
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    sdkTraceBase = require('@opentelemetry/sdk-trace-base');
  } catch {
    sdkTraceBase = null;
  }
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    sdkTraceNode = require('@opentelemetry/sdk-trace-node');
  } catch {
    sdkTraceNode = null;
  }
  if (!sdkTraceBase) {
    getLogger().warn(
      'attachSpanExporter: @opentelemetry/sdk-trace-base is not installed; ' +
        'spans will not be exported. Install ' +
        '@opentelemetry/sdk-trace-base + @opentelemetry/sdk-trace-node.',
    );
    return;
  }

  const api = _tryLoadOtelApi();
  if (!api) return;

  let provider:
    | ({
        addSpanProcessor?(p: unknown): void;
        _patterAttachedExporters?: Set<unknown>;
      } & Record<string, unknown>)
    | null = null;

  try {
    // Prefer the existing global provider — never replace a host-app
    // TracerProvider silently (parity with Python's behaviour).
    const tracerApi = api.trace as unknown as {
      getTracerProvider?(): unknown;
    };
    const existing = tracerApi.getTracerProvider?.() ?? null;
    if (
      existing &&
      typeof (existing as { addSpanProcessor?: unknown }).addSpanProcessor === 'function'
    ) {
      provider = existing as unknown as typeof provider;
    }
  } catch {
    provider = null;
  }

  if (!provider) {
    if (!sdkTraceNode) {
      getLogger().warn(
        'attachSpanExporter: no SDK TracerProvider registered and ' +
          '@opentelemetry/sdk-trace-node is not installed; cannot wire exporter.',
      );
      return;
    }
    try {
      provider = new sdkTraceNode.NodeTracerProvider();
      const trace = api.trace as unknown as {
        setGlobalTracerProvider?(p: unknown): void;
      };
      trace.setGlobalTracerProvider?.(provider);
    } catch (e) {
      getLogger().debug(
        `attachSpanExporter: failed to construct NodeTracerProvider: ${String(
          (e as Error)?.message ?? e,
        )}`,
      );
      return;
    }
  }

  // Idempotency: track attached exporters by reference on the provider.
  let seen = provider._patterAttachedExporters;
  if (!seen) {
    seen = new Set<unknown>();
    provider._patterAttachedExporters = seen;
  }
  if (seen.has(exporter)) return;

  try {
    const processor = new sdkTraceBase.SimpleSpanProcessor(exporter);
    provider.addSpanProcessor?.(processor);
    seen.add(exporter);
  } catch (e) {
    getLogger().debug(
      `attachSpanExporter: failed to register exporter: ${String(
        (e as Error)?.message ?? e,
      )}`,
    );
  }
}

/** Internal: reset module state (primarily for tests; not part of the public API). */
export function _resetPatterAttrsForTesting(): void {
  _scopeStack.length = 0;
}

File: apps/patter-voice-sdk/libraries/python/getpatter/observability/attributes.py
"""patter.* span attribute helpers.

Lazy-OTel-guarded helpers used by ``getpatter`` to stamp ``patter.cost.*``
and ``patter.latency.*`` attributes on spans during a call's lifecycle.
The two ContextVars (``patter.call_id`` and ``patter.side``) propagate
through asyncio task trees so spans emitted by deeply nested provider
code inherit the call's identity automatically.

See ``docs/DEVLOG.md`` for the version decision and rollout history.
"""

from __future__ import annotations

import logging
from contextlib import contextmanager
from contextvars import ContextVar
from typing import Any, Iterator, Mapping

logger = logging.getLogger("getpatter.observability")

try:
    from opentelemetry import trace as _trace

    _OTEL = True
except ImportError:  # pragma: no cover — optional [tracing] extra
    _trace = None  # type: ignore[assignment]
    _OTEL = False

DEFAULT_SIDE = (
    "uut"  # "unit-under-test"; the side value used when no driver/UUT split is in play.
)

_patter_call_id: ContextVar[str | None] = ContextVar("patter.call_id", default=None)
_patter_side: ContextVar[str] = ContextVar("patter.side", default=DEFAULT_SIDE)


def record_patter_attrs(attrs: Mapping[str, Any]) -> None:
    """Stamp ``patter.*`` attributes on the current span, plus call_id and side.

    Behaviour:
    - No-op if OTel is missing or no ``patter_call_scope`` is active.
    - If a recording span is active, attributes are stamped on it.
    - If no recording span is active, a transient zero-duration
      ``patter.billable`` span is opened solely to carry the attributes.
      This is a best-effort fallback for callers without their own span;
      downstream collectors that filter zero-duration spans may drop
      these. Callers that want guaranteed attribution should wrap their
      billable work in their own span.

    Caller-provided ``patter.call_id`` / ``patter.side`` keys win over
    the ContextVar values (via ``setdefault``).
    """
    if not _OTEL:
        return
    call_id = _patter_call_id.get()
    if call_id is None:
        return
    side = _patter_side.get()
    full = dict(attrs)
    # setdefault: caller-provided patter.call_id/side wins
    full.setdefault("patter.call_id", call_id)
    # setdefault: caller-provided patter.call_id/side wins
    full.setdefault("patter.side", side)

    span = _trace.get_current_span()
    if span is not None and span.is_recording():
        for k, v in full.items():
            span.set_attribute(k, v)
        return

    tracer = _trace.get_tracer("getpatter.observability")
    with tracer.start_as_current_span("patter.billable") as new_span:
        for k, v in full.items():
            new_span.set_attribute(k, v)


@contextmanager
def patter_call_scope(*, call_id: str, side: str = DEFAULT_SIDE) -> Iterator[None]:
    """Bind call_id and side to the current asyncio task tree."""
    if not call_id:
        raise ValueError("patter_call_scope requires non-empty call_id")
    cid_token = _patter_call_id.set(call_id)
    side_token = _patter_side.set(side)
    try:
        yield
    finally:
        _patter_call_id.reset(cid_token)
        _patter_side.reset(side_token)


def attach_span_exporter(
    patter_instance: Any, exporter: Any, *, side: str = DEFAULT_SIDE
) -> None:
    """Wire ``exporter`` into the global TracerProvider via SimpleSpanProcessor.

    Stores ``side`` on the Patter instance (``_patter_side`` attr) so the
    per-call handler reads it when entering ``patter_call_scope``.

    Idempotency contract: idempotent on the *same exporter object reference*.
    If the caller constructs two distinct exporter instances pointing at the
    same backend (e.g. two ``OTLPSpanExporter(endpoint=...)`` calls), both
    will be attached and spans will be exported twice. Hold a single
    exporter object and pass it on every call to avoid duplicates.

    If the global TracerProvider is not a ``TracerProvider`` instance
    (e.g. the no-op ``ProxyTracerProvider``), it is replaced with a fresh
    one and a warning is logged.
    """
    patter_instance._patter_side = side

    if not _OTEL:
        logger.debug("attach_span_exporter: OTel not installed; only side= stored")
        return

    try:
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import SimpleSpanProcessor
    except ImportError:
        logger.warning(
            "attach_span_exporter: opentelemetry-sdk not installed; "
            "spans will not be exported. Install getpatter[tracing]."
        )
        return

    provider = _trace.get_tracer_provider()
    if not isinstance(provider, TracerProvider):
        # Replacing a non-SDK provider would destroy any host-app
        # instrumentation already attached. Warn loudly so the operator
        # can pass their own SDK TracerProvider via init_tracing() instead.
        logger.warning(
            "attach_span_exporter: replacing existing TracerProvider %r with "
            "a fresh getpatter-managed TracerProvider. If your host app uses "
            "OTel auto-instrumentation, configure a TracerProvider before "
            "calling _attach_span_exporter to avoid losing those processors.",
            type(provider).__name__,
        )
        provider = TracerProvider()
        _trace.set_tracer_provider(provider)

    seen = getattr(provider, "_patter_attached_exporters", None)
    if seen is None:
        seen = set()
        provider._patter_attached_exporters = seen
    if id(exporter) in seen:
        return
    provider.add_span_processor(SimpleSpanProcessor(exporter))
    seen.add(id(exporter))
